c10/core/impl/cow/state_machine.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <c10/core/impl/cow/shadow_storage.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <cstdint>
+
+namespace c10::impl::cow {
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
+//
+// This type manages the following transition:
+
+// 1) A tensor is created. It has a StateMachine instance that is in
+//    the StateId::initial state. There are no shadow storages. This tensor and
+//    any views created from it will have a null shadow storage.
+//
+// 2) A lazy copy is created, e.g. from Tensor::reshape(). This
+//    transitions from StateId::initializing and then finally to
+//    StateId::active. At this point a StateMachine::Impl has been
+//    allocated on the heap.
+//
+//    The Tensor that was the source of the lazy copy, and any views
+//    created from it, will still have a null shadow storage pointer,
+//    but its actual shadow storage will be in
+//    Impl::default_shadow_storage.
+class C10_API StateMachine {
+ public:
+  /** @see cow::ShadowStorage::Generation */
+  using Generation = cow::ShadowStorage::Generation;
+
+  // Constructs an instance in the "initial" state.
+  StateMachine();
+
+  // Responsible for cleaning up any implementation details.
+  ~StateMachine();
+
+  // Gets the current generation of the physical storage.
```

Comment:
nit: Worth saying that no tracking happens UNTIL we simulate a lazy copy, so you're not allowed to use this to track mutations to the storage otherwise